pkg/controllers/vdb/onlineupgrade_reconciler.go

```diff
@@ -1113,11 +1117,27 @@ func (r *OnlineUpgradeReconciler) promoteSandboxToMainCluster(ctx context.Contex
 		r.Log.Info("Waiting for all pods in sandbox to be up for promotion.", "sandboxName", r.sandboxName)
 		return ctrl.Result{Requeue: true}, nil
 	}
+	anns := map[string]string{
+		vmeta.OnlineUpgradePromotionAttemptAnnotation: strconv.Itoa(vmeta.GetOnlineUpgradePromotionAttempt(r.VDB.Annotations) + 1),
+	}
+	_, err = vk8s.MetaUpdateWithAnnotations(ctx, r.VRec.GetClient(), r.VDB.ExtractNamespacedName(), r.VDB, anns)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
 	actor := MakePromoteSandboxToMainReconciler(r.VRec, r.Log, r.VDB, sbPFacts, r.Dispatcher, r.VRec.Client)
 	r.Manager.traceActorReconcile(actor)
-	res, err := actor.Reconcile(ctx, &ctrl.Request{})
+	var res ctrl.Result
+	res, err = actor.Reconcile(ctx, &ctrl.Request{})
 	if verrors.IsReconcileAborted(res, err) {
-		return res, err
+		attempts := vmeta.GetOnlineUpgradePromotionAttempt(r.VDB.Annotations)
+		if attempts >= vmeta.OnlineUpgradePromotionMaxAttempts {
+			r.Log.Info("The last 3 sandbox promotion attempts have failed. Revive the database with the restorepoint", "sandboxName", r.sandboxName)
+			r.VRec.Eventf(r.VDB, corev1.EventTypeNormal, events.UpgradeFailed,
+				"Online upgrade has failed after 3 sandbox promotion attempts. Use restorepoint in archive %s to revive the database",
```

Comment:
Change "Use restorepoint in archive %s to revive the database" to "Please revive the database to its original state."